### PR TITLE
Add drop_duplicates for dims

### PR DESCRIPTION
**Pull request by ahuang11**
Ruined https://github.com/pydata/xarray/pull/5089 with reverting so remaking the PR for just dims

- [ ] Closes #xxxx
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Review by shoyer — APPROVED** [review]

**Review by shoyer — CHANGES_REQUESTED** [review]
Actually -- there is still one unresolved design concern here.

If you drop duplicates over multiple variables at once, what dimensions should the result have? In particular -- should it have the original dimensions, or should all dimensions involved be combined into one?

The latter might seem a little crazy now, but would make more sense once we allow dropping over multi-dimensional variables.

If we really want to avoid any possible controversy here, it might be best to stick to only supporting one variable for now, too.

**Comment by max-sixty** [comment]
The docs are failing with:

```
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/5239/xarray/core/dataarray.py:docstring of xarray.core.dataarray.DataArray.drop_duplicates:3: WARNING: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/5239/xarray/core/dataarray.py:docstring of xarray.core.dataarray.DataArray.drop_duplicates:4: WARNING: Block quote ends without a blank line; unexpected unindent.
looking for now-outdated files... none found
```

**Comment by max-sixty** [comment]
> If you drop duplicates over multiple variables at once, what dimensions should the result have? In particular -- should it have the original dimensions, or should all dimensions involved be combined into one?
> 
> The latter might seem a little crazy now, but would make more sense once we allow dropping over multi-dimensional variables.
> 
> If we really want to avoid any possible controversy here, it might be best to stick to only supporting one variable for now, too.

For me this case supports consistent behavior over any number of dims! Which would mean rejecting the "stack only if you supply exactly ndims" proposal.

Completely fine with restricting to one dim for the moment and seeing how that goes (it's always possible to pass `.drop_duplicates_coords(dim1).drop_duplicates_coords(dim2)`.

Were we going with `drop_duplicates` or `drop_duplicates_coords`? No strong view from me at all.

File: xarray/core/dataarray.py
Comment (keewis) [inline xarray/core/dataarray.py:4491]:
```suggestion
        dims : hashable or sequence of hashable, optional
```

**Comment by keewis** [comment]
the docs build can probably be fixed by adding newlines before the section headers ("Parameters" / "Returns")

**Comment by max-sixty** [comment]
@ahuang11 we discussed this on the core team call. People are excited to merge this, and appreciate you bearing with the changes.

I suggested that we narrow this even further to one dimension and merge, so we can benefit from this now and consider additions from there. Would that be OK with you?

**Comment by ahuang11** [comment]
Sure.

What's the reasoning for a single dimension?

**Comment by max-sixty** [comment]
@shoyer could I confirm that limiting to a single dimension satisfies this? You say "multiple variables" here. I mentioned a single dimension on the call but maybe wasn't clear.

> If you drop duplicates over multiple variables at once, what dimensions should the result have? In particular -- should it have the original dimensions, or should all dimensions involved be combined into one?
> 
> The latter might seem a little crazy now, but would make more sense once we allow dropping over multi-dimensional variables.
> 
> If we really want to avoid any possible controversy here, it might be best to stick to only supporting one variable for now, too.

**Comment by ahuang11** [comment]
Dont understand this
```
xarray/core/dataarray.py:4605: error: Argument 1 to "isel" of "DataArray" has incompatible type "Dict[str, Any]"; expected "Optional[Mapping[Hashable, Any]]"  [arg-type]
Found 1 error in 1 file (checked 140 source files)
```

`self.isel(indexes)`

File: xarray/core/dataarray.py
Comment (max-sixty) [inline xarray/core/dataarray.py:4581]:
```suggestion
        dim: Hashable,
```

File: xarray/core/dataarray.py
Comment (max-sixty) [inline xarray/core/dataarray.py:4601]:
```suggestion
```

File: xarray/core/dataarray.py
Comment (max-sixty) [inline xarray/core/dataarray.py:4604]:
```suggestion
        indexes = {dim: ~self.get_index(dim).duplicated(keep=keep)}
```

**Comment by dcherian** [comment]
> @shoyer could I confirm that limiting to a single dimension satisfies this? You say "multiple variables" here. I mentioned a single dimension on the call but maybe wasn't clear.

@shoyer can you kindly merge this if you're happy with it?

**Comment by max-sixty** [comment]
Merging but I or @ahuang11 can do any follow-ups.

Thank you v much @ahuang11 — appreciate this was a lot of work and happy to have you contribute such a good feature.

**Pull request merged**
